v2/apiserver/schemas/common.json

```diff
@@ -0,0 +1,76 @@
+{
+	"$schema": "http://json-schema.org/draft-07/schema#",
+	"$id": "github.com/brigadecore/brigade/common.schema.json",
+	"description": "A common set of schema definitions shared by other Brigade schemas",
+
+	"definitions": {
+
+		"apiVersion": {
+			"type": "string",
+			"description": "The major version of the Brigade API with which this object conforms",
+			"enum": ["brigade.sh/v2"]
+		},
+
+		"description": {
+			"type": "string",
+			"minLength": 3,
+			"maxLength": 80
+		},
+
+		"empty": {
+			"type": "string",
+			"enum": [ "" ]
+		},
+
+		"gitConfig": {
+			"type": "object",
+			"description": "Worker configuration pertaining specifically to git",
+			"additionalProperties": false,
+			"properties": {
+				"cloneURL": {
+					"oneOf": [
+						{ "$ref": "#/definitions/empty" },
+						{ "$ref": "#/definitions/url" }
+					],
+					"description": "The URL for cloning a git project"
+				},
+				"commit": {
+					"type": "string",
+					"pattern": "^[a-fA-F0-9]*$",
+					"minLength": 8,
+					"maxLength": 40,
+					"description": "A git commit sha"
+				},
+				"ref": {
+					"type": "string",
+					"description": "A reference to a git branch or tag"
+				},
+				"initSubmodules": {
+					"type": "boolean",
+					"description": "Whether to initialize git submodules"
+				}
+			}
+		},
```

Comment:
This shouldn't be common. When using this in a project definition, there is a an `initSubmodules` field that isn't supposed to be included when using this in an event....

But I just looked, and it _was_ identical in both places-- my fault; not yours.

Can this one section be removed from common can have it be defined differently for projects (include `initSubmodules` field) and events (exclude `initSubmodules`) field?

Reply:
I bet we could reconfigure things to support that.

However, I'm wondering if there might be utility in allowing an event to toggle `initSubmodules` and override a project's default?  We'd need to update the event struct as well, of course.  But wanted to pose that possibility.

Reply:
I don't recall v1 supporting it.

I wouldn't discount the possibility that there's utility in it for some edge case, but as things stand now, this is in the realm of things where the "project admin" (once roles are enforced) can specify permissible behavior. I think it's similar in many respects to the project admin setting the policy for whether jobs are permitted to run as privileged or mount the host's Docker socket-- i.e. a security concern. After all, our support for initializing submodules is an avenue for bringing third-party code that isn't part of a PR into the environment for compilation.

Reply:
Ok... my thinking on this has changed, slightly.

If we don't initialize the submodules, there's nothing that actually stops someone from creating a job that does it, so allowing `initSubmodules` to be specified at the event-level really open a new avenue of attack. Consider my concerns over security moot.

That said, I can't find the utility of `initSubmodules` at the event-level. The practical use for _any_ event-level git details is to enabled CI/CD workflows by identifying specific coordinates (i.e. commit, ref, etc.) for code that is to be tested, built, deployed, etc. It's job is _not_ to say _how_ to build, but only _what_ to build.

Defining _how_ to build is the responsibility of the project definition.

For this reason, I stand by the notion that the `initSubmodules` field only makes sense in the context of the project-level git configuration.

Reply:
I appreciate the thorough consideration, here!  I'm good with keeping things as is (no `initSubmodules` in the event schema's gitConfig) and will see what it looks like to consolidate the rest of the git bits.

Reply:
sgtm

Reply:
Updated, PTAL when convenient.